### PR TITLE
Enable vectorized dim repeat.

### PR DESCRIPTION
**Pull request by kimishpatel**
Summary:
This diff enables vectorized repeat op for contigous  tensors via broadcast and
memcpy instead of repeat's elementwise copy.

For now this is introduced as a separate op. I plan to integrate this in repeat op in the next PR unless we want to do this from the get-go.

Bandwidth calculation was done as follows:
Num_bytes = input_tensor.nbytes() + output_tensor.nbytes() (This is how many bytes we are reading and writing)
Time_ms = amount of time taken during repeat op.
MB_in_bytes = (1/(1000 * 1000))
Ms_in_second = 1000

BW = (bytes * MB_in_bytes) / (time_ms / ms_in_second))

| Input Tensor	| Repeat | Output Size	| Existing BW (MB/s) | Vectorized Copy BW (MB/s) | MemCpy | Vectorized Copy/Existing |
| ---- | ----| --- | --- | --- | --- | --- |
| [1, 64, 128, 32] | [128, 1, 1, 1]|671088640 | 758 | 1964 | 1509 |2.59 |
| [32, 32, 256, 1] |[1, 1, 1, 128]|33554432 | 31 | 1928 | 1882 |62.19 |
| [256, 256, 16, 8] |[1, 1, 16, 1]|134217728 | 371 | 2120 | 934 |5.71 |
| [512, 512] |[1, 16]|4194304 | 820 | 2006 | 1410 | 2.45 |
| [8, 256, 128]|[32, 1, 1]|8388608 | 832| 2182 | 1560 | 2.62 |
| [128, 256, 1]|[1, 1, 32]|1048576 | 408 | 4388 | 3943 | 10.75 |

Differential Revision: D14935926

### CODE REVIEW

File: aten/src/ATen/native/cpu/VectorizedCopy.h [hunk 1]
```diff
@@ -0,0 +1,108 @@
+#pragma once
```
Comment (jamesr66a) [inline aten/src/ATen/native/cpu/VectorizedCopy.h]: So actually the compilation unit(s) that this is included in need to be in `native/cpu`, not just this header. Probably need to split out the implementation from TensorShape.cpp and use dynamic dispatch

File: aten/src/ATen/native/cpu/VectorizedCopy.h [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+#include <ATen/ATen.h>
+#include <ATen/cpu/vec256/vec256.h>
+
+namespace at {
+namespace native {
+
+template <typename T>
+using VecType = at::vec256::Vec256<T>;
```
Comment (jamesr66a) [inline aten/src/ATen/native/cpu/VectorizedCopy.h]: Is this unused?

File: aten/src/ATen/native/cpu/VectorizedCopy.h [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    dest_ptr += size;
+  }
+  for (uint64_t j = 0; j < repeat_factor % size; ++j) {
+    *dest_ptr = src_element;
+    ++dest_ptr;
+  }
+}
+
+template <typename DataType>
+void vectorized_copy(DataType* dest_ptr, const DataType* src_ptr, const uint64_t num_elements) {
```
Comment (jamesr66a) [inline aten/src/ATen/native/cpu/VectorizedCopy.h]: This function seems unnecessary

**Comment by jamesr66a** [comment]
BTW this is still underutilizing the memory bandwidth. In my short investigation it seems that expand() and alias() cause heavy memory allocation, and the cost of that is more than the data movement itself. Might be something to investigate

**Comment by dzhulgakov** [comment]
I think we should integrate it in the corresponding op from the get go. There's not good reason whatsoever to expose vectorized_contig_tensor_repeat to user. I suppose you can pattern-match the strides in the implementation and call the fast version, right?

**Comment by cpuhrsch** [comment]
Instead of creating a new function we should branch within the implementation of repeat. And as James already pointed out, in order to take advantage of vectorization on the CPU, the file needs to be written in a kernel. You can use the UnaryOps.* files as guidance. cc @VitalyFedyunin as the new CPU POC.

File: aten/src/ATen/native/cpu/VectorizedCopy.h (hunk 1 above)
Comment (kimishpatel) [reply aten/src/ATen/native/cpu/VectorizedCopy.h]: Actually, I will just remove this as exposed API and integrate it with repeat.

File: aten/src/ATen/native/cpu/VectorizedCopy.h (hunk 2 above)
Comment (kimishpatel) [reply aten/src/ATen/native/cpu/VectorizedCopy.h]: Yup. Will remove.

File: aten/src/ATen/native/cpu/VectorizedCopy.h (hunk 3 above)
Comment (kimishpatel) [reply aten/src/ATen/native/cpu/VectorizedCopy.h]: Sure.

**Comment by kimishpatel** [comment]
@jamesr66a where do you see the memory allocation being the larger portion of the op? Can you share your setup? In the doc I shared I profiled vectorized repeat and it seems to spend most of the time doing copying. There is some memory allocation for the output tensor but that does not show up in where most of the time is spent.

**Comment by kimishpatel** [comment]
@cpuhrsch by exposing the API it helped test the implementation against default repeat. What tests can I run, if I remove API and integrate into repeat, to make sure I did not break anything.

**Comment by cpuhrsch** [comment]
@kimishpatel - We advice to a) check the coverage of the currently existing tests by confirming they exercise your branch b) extend the tests if necessary to explicitly exercise your work.

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -1461,6 +1461,9 @@
   dispatch:
     CPU: _neg_out_cpu
     CUDA: _neg_out_cuda
+- func: vectorized_contig_tensor_repeat(Tensor self, int[] repeats) -> Tensor
+  matches_jit_signature: True
```
Comment (cpuhrsch) [inline aten/src/ATen/native/native_functions.yaml]: This is true by default and not necessary to add anymore, if the signature is compliant.

**Comment by kimishpatel** [comment]
@cpuhrsch what are the existing tests? test/test_torch?

File: aten/src/ATen/native/cpu/VectorizedCopy.h (hunk 1 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/VectorizedCopy.h]: Is there a reason all of this lives in a header?

File: aten/src/ATen/native/cpu/VectorizedCopy.h (hunk 3 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/VectorizedCopy.h]: A vectorized copy is being worked on separately. cc @VitalyFedyunin

**Comment by cpuhrsch** [comment]
@kimishpatel - torch_test should include this, but it's also used from torch.nn etc. - Typically the whole test suite exercises our ops from various perspectives. 

An explicit test should be part of test_repeat within test_torch. You can also look for explicit invocations of repeat within the tests.

File: aten/src/ATen/native/cpu/VectorizedCopy.h (hunk 1 above)
Comment (kimishpatel) [reply aten/src/ATen/native/cpu/VectorizedCopy.h]: @cpuhrsch Better code refactoring? I am open to suggestions.

Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/VectorizedCopy.h]: We've been writings these kernels within .cpp files. Take a look at UnaryKernels.cpp. This also includes information on how to setup the dispatch etc.

**Comment by jamesr66a** [comment]
@kimishpatel this is the hottest callstack on my mac:

image

perf on devserver doesn't show deep enough stacks for some reason, but calls to `malloc` are similarly expensive

File: aten/src/ATen/native/cpu/VectorizedCopy.h (hunk 1 above)
Comment (kimishpatel) [reply aten/src/ATen/native/cpu/VectorizedCopy.h]: @cpuhrsch, I am gonna move this file out of cpu folder and into native for now. I will multiple dispatch option thin in a separate PR.

**Comment by colesbury** [comment]
Is this the same op as Tensor.repeat?

**Comment by kimishpatel** [comment]
@colesbury yes it is. I have upcoming update to this PR that just add a branch in repeat to call the vectorized version under certain conditions.

**Comment by colesbury** [comment]
@kimishpatel -- I'm not sure it makes sense to continue working on this. I think this will be completely handled by an efficient copy implementation, which should land very soon.

**Comment by kimishpatel** [comment]
@colesbury can you point me to the PR?

**Comment by colesbury** [comment]
@kimishpatel https://github.com/pytorch/pytorch/pull/18618

**Comment by soumith** [comment]
I think @colesbury is referring to @VitalyFedyunin 's PR https://github.com/pytorch/pytorch/pull/18618

**Comment by kimishpatel** [comment]
@colesbury Note that this PR has two specialization. 1) When the repeat is across last time with one element. In that case it uses broadcast and vector store. Note sure if the PR #18618 will let compiler see through that. 2) It uses memcpy instead of for all other cases of contig tensors instead of manually vectorizing them via vec256. I found that to perform better. 
If you still think that this PR is redundant I am ok with it, but maybe we can benchmark the bandwidth achieved for different tensor sizes for repeat.

**Comment by VitalyFedyunin** [comment]
Hello! How about: I finish updating copy and make it support repeating. We compare performance of both implementations and check if we need separate (faster) implementation of dim repeat.

**Comment by kimishpatel** [comment]
@VitalyFedyunin sounds good to me.

**Comment by kimishpatel** [comment]
@VitalyFedyunin any update on this?

**Comment by VitalyFedyunin** [comment]
It is pretty much ready and going to be landed after benchmarks is done #18618 
I'm not going to add special function, `copy_` with the broadcasting should do the trick.

**Comment by kimishpatel** [comment]
@VitalyFedyunin, let me know once you land it. I will check the impact of this on repeat in my benchmarking.

**Comment by VitalyFedyunin** [comment]
I think we can close this one, right?
